### PR TITLE
feat(bitbucket): support task autocomplete

### PR DESCRIPTION
**Pull request by dandandy**
## Changes

Add a new renovate option 'bbAutoResolvePrTasks' which when set to true (the default is false) will mark PR tasks as "resolved" after the PR is created. This only works in bitbucket cloud. 

## Context

Closes https://github.com/renovatebot/renovate/discussions/27485#discussioncomment-10009078

This feature is useful for repos which set default tasks because it:
- saves reviewers from having to click on the checkbox to manually resolves tasks
- unblocks automerging (currently automerge fails if there are unresolved tasks)

## Documentation (please check one with an [x])

- [ x ] I have updated the documentation

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ x ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

I plan to add unit tests, but I wanted to raise this first to get any inital feedback before I write tests (as per your contribution docs).

### CODE REVIEW

**Comment by dandandy** [comment]
Here is an excerpt from the DEBUG logs when I ran it locally against my branch with config:

renovate.json file
```json

```

run command:
```bash
LOG_LEVEL=debug RENOVATE_CONFIG_FILE=config.js node  ../renovate/dist/renovate.js
```

config.js
```javascript
module.exports = {
    endpoint: 'https://api.bitbucket.org/',
      repositories: ['testing-dbarrett/testing-repo'],
    platform: 'bitbucket',
    username: '***redacted***'
}
```

```
DEBUG: Creating PR (repository=testing-dbarrett/testing-repo, branch=renovate/react-scripts-3.x)
       "prTitle": "Update dependency react-scripts to ^3.4.4"
DEBUG: Creating PR (repository=testing-dbarrett/testing-repo, branch=renovate/react-scripts-3.x)
       "title": "Update dependency react-scripts to ^3.4.4",
       "base": "main"
DEBUG: Using queue: host=api.bitbucket.org, concurrency=16 (repository=testing-dbarrett/testing-repo, branch=renovate/react-scripts-3.x)
DEBUG: Auto resolve PR tasks (repository=testing-dbarrett/testing-repo, branch=renovate/react-scripts-3.x)
       "title": "Update dependency react-scripts to ^3.4.4",
       "base": "main",
       "bbAutoResolvePrTasks": true
DEBUG: List PR tasks (repository=testing-dbarrett/testing-repo, branch=renovate/react-scripts-3.x)
       "title": "Update dependency react-scripts to ^3.4.4",
       "base": "main",
       "bbAutoCompleteTasks": true,
       "listTaskRes": [
         {
           "id": 32224203,
           "state": "UNRESOLVED",
           "content": "[content]",
           "creator": null,
           "created_on": "2024-08-20T01:42:20.646432+00:00",
           "updated_on": "2024-08-20T01:42:20.646497+00:00",
           "resolved_on": null,
           "resolved_by": null,
           "pending": false,
           "links": {
             "self": {
               "href": "https://api.bitbucket.org/2.0/repositories/testing-dbarrett/testing-repo/pullrequests/14/tasks/32224203"
             },
             "html": {
               "href": "https://bitbucket.org/testing-dbarrett/testing-repo/pull-requests/14/_/diff#task-32224203"
             }
           }
         }
       ]
DEBUG: Put PR tasks - mark resolved (repository=testing-dbarrett/testing-repo, branch=renovate/react-scripts-3.x)
       "title": "Update dependency react-scripts to ^3.4.4",
       "base": "main",
       "updateTaskResponse": {
         "statusCode": 200,
         "body": {
           "id": 32224203,
           "state": "RESOLVED",
           "content": "[content]",
           "creator": null,
           "created_on": "2024-08-20T01:42:20.646432+00:00",
           "updated_on": "2024-08-20T01:42:21.658921+00:00",
           "resolved_on": "2024-08-20T01:42:21.658916+00:00",
           "resolved_by": {
             "display_name": "Daniel Barrett",
             "links": {
               "self": {
                 "href": "https://api.bitbucket.org/2.0/users/%7Be0356302-f882-4657-bf0e-63fa22579be3%7D"
               },
               "avatar": {
                 "href": "https://avatar-management--avatars.us-west-2.prod.public.atl-paas.net/initials/DB-1.png"
               },
               "html": {
                 "href": "https://bitbucket.org/%7Be0356302-f882-4657-bf0e-63fa22579be3%7D/"
               }
             },
             "type": "user",
             "uuid": "{e0356302-f882-4657-bf0e-63fa22579be3}",
             "account_id": "70121:29895d36-501b-4744-93ed-25484d8556a6",
             "nickname": "Daniel Barrett"
           },
           "pending": false,
           "links": {
             "self": {
               "href": "https://api.bitbucket.org/2.0/repositories/testing-dbarrett/testing-repo/pullrequests/14/tasks/32224203"
             },
             "html": {
               "href": "https://bitbucket.org/testing-dbarrett/testing-repo/pull-requests/14/_/diff#task-32224203"
             }
           }
         },
         "headers": {
           "date": "Tue, 20 Aug 2024 01:42:21 GMT",
           "content-type": "application/json; charset=utf-8",
           "server": "AtlassianEdge",
           "vary": "Authorization, Origin, cookie, user-context",
           "x-accepted-oauth-scopes": "pullrequest",
           "x-used-mesh": "False",
           "x-credential-type": "apppassword",
           "x-token-id": "4741201",
           "x-oauth-scopes": "issue:write, pullrequest:write, team, account",
           "x-view-name": "bitbucket.apps.pullrequests.api.v20.handlers.PullRequestTaskHandler",
           "x-dc-location": "Micros-3",
           "x-served-by": "e1428dd1c9ea",
           "x-version": "19e939a229ff",
           "x-static-version": "19e939a229ff",
           "x-request-count": "200",
           "x-render-time": "0.22031736373901367",
           "x-b3-traceid": "f312c246e5b74aa5841846f0c1eb430a",
           "x-b3-spanid": "1b21104d5cb763ec",
           "x-frame-options": "SAMEORIGIN",
           "expires": "Tue, 20 Aug 2024 01:42:21 GMT",
           "cache-control": "max-age=0, no-cache, no-store, must-revalidate, private",
           "x-usage-user-time": "0.112488",
           "x-usage-system-time": "0.005170",
           "x-usage-input-ops": "0",
           "x-usage-output-ops": "0",
           "x-trace-id": "f312c246e5b74aa5841846f0c1eb430a",
           "strict-transport-security": "max-age=31536000; includeSubDomains; preload",
           "content-encoding": "gzip",
           "x-content-type-options": "nosniff",
           "x-xss-protection": "1; mode=block",
           "atl-traceid": "f312c246e5b74aa5841846f0c1eb430a",
           "report-to": "{\"endpoints\": [{\"url\": \"https://dz8aopenkvv6s.cloudfront.net\"}], \"group\": \"endpoint-1\", \"include_subdomains\": true, \"max_age\": 600}",
           "nel": "{\"failure_fraction\": 0.001, \"include_subdomains\": true, \"max_age\": 600, \"report_to\": \"endpoint-1\"}",
           "transfer-encoding": "chunked"
         },
         "authorization": "***********"
       }
 INFO: PR created (repository=testing-dbarrett/testing-repo, branch=renovate/react-scripts-3.x)
       "pr": 14,
       "prTitle": "Update dependency react-scripts to ^3.4.4"
```

**Comment by rarkins** [comment]
@dandandy please complete the CLA and mark as ready for review when you want us to look at it

File: lib/modules/platform/bitbucket/index.ts
Comment (rarkins) [inline lib/modules/platform/bitbucket/index.ts:967]: log only if it's true. Also you don't need to log the repository as it's already logged in metadata

File: lib/modules/platform/bitbucket/index.ts
Comment (rarkins) [inline lib/modules/platform/bitbucket/index.ts:968]: Is there any scenario where `pr` is null or undefined at this point?

File: lib/modules/platform/bitbucket/index.ts
Comment (rarkins) [inline lib/modules/platform/bitbucket/index.ts:969]: Code below here to be wrapped in try/catch so that a tasks completion problem doesn't blow up the entire createPr() method

File: lib/modules/platform/bitbucket/index.ts
Comment (rarkins) [inline lib/modules/platform/bitbucket/index.ts:1001]: This is too verbose
```suggestion
      logger.trace(
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/platform/bitbucket/index.ts
Comment (dandandy) [reply lib/modules/platform/bitbucket/index.ts]: Looking at the code paths there shouldn't be a scenario where it's null, but to be defensive I've added a null check to the if statement

Comment (dandandy) [reply lib/modules/platform/bitbucket/index.ts]: I think it's a bit cleaner/simpler to keep `pr` as a  const, and move the `bbAutoResolvePrTasks` logic into a separate function, and call that in the appropriate places. I've added this as a separate commit. Let me know what you think.

File: lib/modules/platform/bitbucket/index.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:987]:
```suggestion
    logger.trace(
```

File: lib/modules/platform/bitbucket/types.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/types.ts:107]: start using zod schema validation

File: lib/modules/platform/bitbucket/index.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:990]: I'll prefer to only log PR Id and drop title and base

File: lib/modules/platform/bitbucket/types.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/types.ts:114]: move to `./schema.ts` and don't use underscore for variables

File: lib/modules/platform/bitbucket/types.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/types.ts:96]:
```suggestion
const TaskState = z.union([z.literal('RESOLVED'), z.literal('UNRESOLVED')]);
```

File: lib/modules/platform/bitbucket/types.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/types.ts:98]:
```suggestion
const PrTask = z.object({
```

File: lib/modules/platform/bitbucket/types.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/types.ts:100]:
```suggestion
  state: TaskState,
```

File: lib/modules/platform/bitbucket/types.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/types.ts:106]:
```suggestion
export type PrTask = z.infer<typeof PrTask>;
```

File: lib/modules/platform/bitbucket/types.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/types.ts:110]:
```suggestion
    values: z.array(PrTask),
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/platform/bitbucket/index.ts
Comment (rarkins) [inline lib/modules/platform/bitbucket/index.ts:973]:
```suggestion
  logger.debug(`Auto resolve PR tasks in #${pr.number}`);
```

File: lib/modules/platform/bitbucket/index.ts
Comment (rarkins) [inline lib/modules/platform/bitbucket/index.ts:1013]:
```suggestion
    logger.warn({ prId: pr.number, err }, 'Error resolving PR tasks');
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Have you tested this against a public bitbucket repo?

**Comment by dandandy** [comment]
@rarkins No I haven't tested against a public repo, just a private one. I can test it against a public one if you want?

**Comment by rarkins** [comment]
@dandandy it would be best if you could, because I'd like to take a look before merging

**Comment by dandandy** [comment]
@rarkins here you go - https://bitbucket.org/gmail-testing/testing-public/pull-requests/2 

I ran renovate locally using a build from commit hash `be8be4d051dfded3d401339f29c402b0f2ad47b1` , renovate is using a app password from my user account, so that's why it shows that I created the Pr and resolved the task. 

Here's an excerpt from the logs:

```
DEBUG: Creating PR (repository=gmail-testing/testing-public, branch=renovate/react-scripts-3.x)
       "prTitle": "Update dependency react-scripts to ^3.4.4"
DEBUG: Creating PR (repository=gmail-testing/testing-public, branch=renovate/react-scripts-3.x)
       "title": "Update dependency react-scripts to ^3.4.4",
       "base": "main"
DEBUG: Auto resolve PR tasks (repository=gmail-testing/testing-public, branch=renovate/react-scripts-3.x)
       "prId": 2
 INFO: PR created (repository=gmail-testing/testing-public, branch=renovate/react-scripts-3.x)
       "pr": 2,
       "prTitle": "Update dependency react-scripts to ^3.4.4"
```

**Review by rarkins — DISMISSED** [review]

**Comment by dandandy** [comment]
Hey @viceice is there anything else you would like me to do for this PR? 

I understand if you're waiting for the right time to merge this in for a release. checking in just in case 😀 

Thanks!

File: lib/modules/platform/bitbucket/index.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:985]:
```suggestion
    const unResolvedTasks = UnresolvedPrTasks.parse(response);
```
pass this type to http request as schema, so validation is done in http layer

File: lib/modules/platform/bitbucket/schema.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/schema.ts:78]:
```suggestion
const TaskState = z.union([z.literal('RESOLVED'), z.literal('UNRESOLVED')]);
```
we use an starting uppercase letter for schemas. typescript can handle this

File: lib/modules/platform/bitbucket/schema.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/schema.ts:80]:
```suggestion
const PrTask = z.object({
```

File: lib/modules/platform/bitbucket/schema.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/schema.ts:82]:
```suggestion
  state: TaskState,
```

File: lib/modules/platform/bitbucket/schema.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/schema.ts:88]:
```suggestion
export type PrTask = z.infer<typeof PrTask>;
```

File: lib/modules/platform/bitbucket/schema.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/schema.ts:92]:
```suggestion
    values: z.array(PrTask),
```

File: lib/modules/platform/bitbucket/schema.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/schema.ts:96]:
```suggestion
    (data) =>
      data.values.filter((task) => task.state === 'UNRESOLVED'),
```
typescript knows the types

**Comment by dandandy** [comment]
hey @viceice @rarkins this pr is ready to be reviewed again.

**Comment by rarkins** [comment]
Please just use the "Request reviewers" UI in github only instead of @ mentions otherwise we get too many notifications

File: lib/modules/platform/bitbucket/index.ts
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:993]: do you really need to send the contents again? 🤔

Comment (dandandy) [reply lib/modules/platform/bitbucket/index.ts:993]: Yep, it's marked as required https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pullrequests/#api-repositories-workspace-repo-slug-pullrequests-pull-request-id-tasks-task-id-put

image

**Comment by dandandy** [comment]
> Please just use the "Request reviewers" UI in github only instead of @ mentions otherwise we get too many notifications

Sorry about that!

**Comment by dandandy** [comment]
Hello, just checking if there's anything else you need from me for this PR? Thanks!

**Review by viceice — APPROVED** [review]

**Pull request merged**
